Chia-Network/chiapos
# Allow limiting number of plotting processes during Stage 1

This PR is based on unmerged https://github.com/Chia-Network/chiapos/pull/176
Please review only the last commit, I'll rebase once 176 is merged. Can be merged separately if needed.
Will require adding commandline switches in chia-blockchain as well.


Sample use case: if you're going to run 16 plotters (each using 4 threads) on 16 thread CPU, it's best to stagger their runs to not lock the computer down during stage 1 (multithreading is supported only in stage 1).

Currently this has to be done with external script:
1. Measure approximate stage 1 time
2. Run 4 plotters
3. Sleep measured time
4. Run next 4 plotters
5. Sleep measured time
etc

But this setup is error-prone:

-  on any slowdown plotter groups will clash
- It's not possible to restart crashed plotters because all timings will be off

As a solution, this patch uses lock files to allow only up to a specified number of processes in Stage 1. Other plotters will wait for their turn. Plotters can be stopped and restarted at any time without consequences. Locks (slots) are automatically released if plotters crash or are stopped.

How to test:

```
./ProofOfSpace create -k 22 --p1maxproc 2 --runtimedir . -d dest1
./ProofOfSpace create -k 22 --p1maxproc 2 --runtimedir . -d dest2
./ProofOfSpace create -k 22 --p1maxproc 2 --runtimedir . -d dest3
```
Runtimedir should be the same (on prod can default to `~/.config/chia/run`)
Third process will wait until first two leave Stage 1.
